File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
class Solution {
    int answer = 0;
    public int solution(int m, int n, String[] board) {
        
        char[][] c = new char[m][n];
        boolean[][] deleted = new boolean[m][n];
        for(int i = 0; i < m; i++) {
            c[i] = board[i].toCharArray();
        }
        while(true) {
            if(!checkBlock(c, deleted)) {
                break;
            }
            deletedBlock(c,deleted);
            dropBlock(c);
        }

        for(char[] row : c) {
            for(char col : row) {
                System.out.print(col + " ");
            }
            System.out.println();
        }
        
        return answer;
    }
    boolean checkBlock(char[][] c, boolean[][] deleted) {
        boolean check = false;
        for(int i = 0; i < c.length-1; i++) {
            for(int j = 0; j < c[i].length-1; j++) {
                char ch = c[i][j];
                if(ch != '-' && c[i][j+1] == ch && c[i+1][j] == ch
                  && c[i+1][j+1] == ch) {
                    check = true;
                    deleted[i][j] = true;
                    deleted[i+1][j] = true;
                    deleted[i][j+1] = true;
                    deleted[i+1][j+1] = true;
                }
            }
        }
        return check;
    }
    void deletedBlock(char[][] c, boolean[][] deleted) {
        for(int i = 0; i < c.length; i++) {
            for(int j = 0; j < c[i].length; j++) {
                if(deleted[i][j]) {
                    c[i][j] = '-';
                    answer++;
                    deleted[i][j] = false;
                }
            }
        }
    }
    void dropBlock(char[][] c) {
        for(int i = 0; i < c.length; i++) {
            for(int j = 0; j < c[i].length; j++) {
                if(c[i][j] == '-') {
                    for(int k = 0; k < i; k++) {
                        if(c[k][j] != '-') {
                            char tmp = c[k][j];
                            c[k][j] = c[i][j];
                            c[i][j] = tmp;
                        }
                    }
                }
            }
        }
    }
}

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.js
function matchBlock(m,n,board) {
    const set = new Set();
    for (let i = 0; i < m-1; i++) {
        for (let j = 0; j < n-1; j++) {
            if(board[i][j] !== '0' && board[i][j] === board[i][j+1] && board[i][j+1] === board[i+1][j] && board[i+1][j] === board[i+1][j+1]) {
                set.add(i+','+j);
                set.add((i+1)+','+j);
                set.add(i+','+(j+1));
                set.add((i+1)+','+(j+1));
            }
        }
    }
    clearBoard(m,n,set,board);
    return set.size;
}

function clearBoard(m,n,set,board) {
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if(set.has(i+','+j)) {
                board[i][j] = '0';
            }
        }
    }
}

function downBlock(m,n,board) {
    for (let j = 0; j < n; j++) {
        const stack = [];
        
        for (let i = 0; i < m; i++) {
            if(board[i][j] !== '0') stack.push(board[i][j]);
        }
        
        for (let i = m - 1; i >= 0; i--) {
            board[i][j] = stack.length > 0 ? stack.pop() : '0';
        }
    }
}

function solution(m, n, board) {
    var answer = 0;
    board = board.map((b) => b.split(''));
    while(true) {
        const size = matchBlock(m,n,board);
        if (size === 0) break;
        answer += size;
        downBlock(m,n,board); 
    }
       
    return answer;
}